utilities/write_batch_with_index/write_batch_with_index.cc

```diff
@@ -61,18 +61,17 @@ class BaseDeltaIterator : public Iterator {
     progress_ = Progress::SEEK_TO_LAST;
 
     // is there an upper bound constraint?
-    if (read_options_ != nullptr &&
-        read_options_->iterate_upper_bound != nullptr) {
+    const Slice* base_upper_bound = base_iterator_upper_bound();
+    if (base_upper_bound != nullptr) {
       // yes, and is base_iterator already constrained by an upper_bound?
-      if (!base_iterator_->has_upper_bound()) {
-        // no, so we have to seek it to before iterate_upper_bound
-        base_iterator_->Seek(*(read_options_->iterate_upper_bound));
+      if (!base_iterator_->check_upper_bound()) {
+        // no, so we have to seek it to before base_upper_bound
+        base_iterator_->Seek(*(base_upper_bound));
```

Comment:
I think the case/code for base_iterator_ and delta_iterator_ should be the same.  Seek to the read_options->upper_bound.  If the iterator is not valid, SeekToLast, otherwise Prev.

I do not think the new methods are required

Reply:
I am not sure I follow you here... the code for `base_iterator_` and `delta_iterator_` are the same apart from the case where the `base_iterator_` might already handle upper/lower bounds.
I did look at making `delta_iterator_` also handle the upper/lower bounds directly itself, but it was too difficult as it is a very different type of iterator which is really just concerned with the SkipList really.

I am fairly certain I have the logic correct as I kept adding tests to ensure this. However, of course it is indeed possible that I missed something... can you see a test case where the logic doesn't hold?

Perhaps it is easier to have a call to discuss?